test/integration/controllermanager/exposureclass/exposureclass_test.go

```diff
@@ -29,21 +28,14 @@ import (
 
 var _ = Describe("ExposureClass controller test", func() {
 	var (
-		resourceName string
-		objectKey    client.ObjectKey
-
 		exposureClass *gardencorev1alpha1.ExposureClass
 		shoot         *gardencorev1beta1.Shoot
 	)
 
 	BeforeEach(func() {
-		resourceName = "test-" + utils.ComputeSHA256Hex([]byte(CurrentSpecReport().LeafNodeLocation.String()))[:8]
-		objectKey = client.ObjectKey{Namespace: testNamespace.Name, Name: resourceName}
-
 		exposureClass = &gardencorev1alpha1.ExposureClass{
 			ObjectMeta: metav1.ObjectMeta{
-				Name:      resourceName,
-				Namespace: testNamespace.Name,
+				GenerateName: testID + "-",
```

Comment:
With this, line 58 will be broken: https://github.com/gardener/gardener/blob/fca73ec9b15e25a86f9a4de1a66aa062c2bf0aaf/test/integration/controllermanager/exposureclass/exposureclass_test.go#L58
`exposureClass.Name` is only populated once the object is created.

This means, there was never any shoot referencing the ExposureClass.

Considering that the test still passed despite this problem, you need to rework this case to test the right thing: 
https://github.com/gardener/gardener/blob/fca73ec9b15e25a86f9a4de1a66aa062c2bf0aaf/test/integration/controllermanager/exposureclass/exposureclass_test.go#L138-L143

Reply:
While debugging the integration test and keeping a breakpoint at Line95 :
https://github.com/gardener/gardener/blob/fca73ec9b15e25a86f9a4de1a66aa062c2bf0aaf/test/integration/controllermanager/exposureclass/exposureclass_test.go#L95

We found out that the `shoot.Spec.ExposureClassName` is having the  correct `exposureClass.Name` value. This might be because  `shoot.Spec.ExposureClassName` is a pointer. WDYT ?

Reply:
Nice finding! You're right, that's why the test works because it is actually correct 😄 

However, this is super implicit and error-prone.
I don't think there is any guarantee that the client decodes into the same field in memory, so you're basically making unsafe assumptions here.
Also, different objects should not share memory, which would have unwanted side effects during decoding.

So, in order to make this safe and explicit, I propose to drop line 58 and add the following after line 93
```go
shoot.Spec.ExposureClassName = pointer.String(exposureClass.Name)
```

Reply:
Done in [commit](https://github.com/gardener/gardener/pull/6423/commits/fca73ec9b15e25a86f9a4de1a66aa062c2bf0aaf).